modules/ml/src/main/java/org/apache/ignite/ml/math/primitives/vector/storage/DenseVectorStorage.java

```diff
@@ -56,19 +61,90 @@ public DenseVectorStorage(double[] data) {
         this.data = data;
     }
 
+    /**
+     * @param data Backing data array.
+     */
+    public DenseVectorStorage(Serializable[] data) {
+        assert data != null;
+
+        this.rawData = data;
+    }
+
     /** {@inheritDoc} */
     @Override public int size() {
-        return data == null ? 0 : data.length;
+        if (data == null && rawData == null)
+            return 0;
+        else {
+            if (data != null)
+                return data.length;
+            if (rawData != null)
+                return rawData.length;
+            else
+                throw new IllegalStateException();
+        }
+    }
+
+    /**
+     * Copies values between internal arrays If toNumericRepresentation = true then method tries cast all serializable
+     * objects to double, initializes double array with converted values and cleans raw data array. If
+     * toNumericRepresentation = false then method copies double array to raw data array and cleans double array.
+     *
+     * @param toNumericRepresentation To numeric representation.
+     */
+    private void swap(boolean toNumericRepresentation) {
+        A.ensure(data == null || rawData == null, "data == null || rawData == null");
+        if (data == null && rawData == null)
+            return;
+
+        if (toNumericRepresentation) {
+            if (data != null)
+                return;
+
+            data = new double[rawData.length];
+            for (int i = 0; i < rawData.length; i++)
+                data[i] = ((Number)rawData[i]).doubleValue();
+            rawData = null;
+        } else {
+            if (rawData != null)
+                return;
+
+            rawData = new Serializable[data.length];
+            for (int i = 0; i < rawData.length; i++)
+                rawData[i] = data[i];
+            data = null;
+        }
     }
 
     /** {@inheritDoc} */
     @Override public double get(int i) {
-        return data[i];
+        if (data != null)
+            return data[i];
+
+        Serializable v = rawData[i];
+        if (v == null)
+            return 0.0;
```

Comment:
We use `Double.NaN` for missing values. Would it be better to use it here?

Reply:
No, it wasn't documented but old API returned 0.0 and I don't want to change this behavior.

Reply:
hmmm, support for Double.NaN but not in this ticket, could you make TODO + ticket for this, @avplatonov ?

Reply:
Yes, sure. But I think we should provide a way to use other user defined values for default values in Vector and use Double.NaN as default way in our code.